### ISSUE TITLE
Network Security (2)

### ISSUE THREAD

**Issue by MapleFall526** [MapleFall526/blog#3]
## ARP Attacks
_Goal_ – sniff all traffic between two hosts in a switched environment.
The attack leverages the stateless nature of the ARP protocol.

### ARP spoofing
The attacker host sends spoofed ARP messages to the two victim hosts, poisoning their ARP cache.
The attacker host acts as a router.
Used in denial of service (DoS) and man-in-the-middle (MITM) attacks.

### MITM exploitations
Variation on the MITM attack:
(1) Use ARP to impersonate the gateway and filter all the traffic to external networks.
(2) Variation of the variation – use ARP to map gateway IP to non-existence MAC address (denial-of-service).

### Tools
- **Dsniff** is a collection of tools for network auditing and penetration testing.
- **Arp** (Linux) – Used to manipulate the system ARP cache.
- **ifconfig** (Linux) – Used to configure/monitor a network interface.
- **Ettercap** – Perform man-in-the-middle attacks in LANs.
   Provide support for
   - ARP spoofing attacks
   - the interception of SSH1 and SSL connections
   - the collection of passwords for a number of protocols.

### ARP Defences
- Static ARP entries
   Configure the ARP cache to ignore dynamic updates.
   Difficult to manage in large installation, could be used for a subset of critical addresses.
- Cache poisoning resistance
   Ignore unsolicited ARP replies, but still vulnerable to hijacking.
   Update on timeout (limited usefulness).
- Monitor changes (**arpwatch**)
   Listen for ARP packets on a local Ethernet interface.
   Keep track for Ethernet/IP address pairs.
   Report suspicious activity and changes in mapping.

### Libnet
Provides a platform-independent library of functions to build arbitrary packets.
Steps in building a packet
(1) Memory Initialisation (allocates memory for packets).
(2) Network Initialisation (initialises the network interface).
(3) Packet Construction (fill in the different protocol headers/payloads).
(4) Packet Injection (send the packet on the wire).

### Blind IP Spoofing
1. A host sends an IP datagram with the address of some other host as the source address.
2. The attacked host replies to the impersonated host.
3. Usually the attacker does not have access to the reply traffic.

### Man-in-the-middle Attacks
An attacker that has control of a gateway used in the delivery process can take advantage of:
- Sniff the traffic
- Intercept, block or delay traffic
- Modify traffic

### Ping of Death
The offset of the last segment is such that the total size of the reassembled datagram is bigger than the maximum allowed size of the receiving kernel. 最后一段的偏移量使重新组装的数据报的总大小大于接收内核的最大允许大小
	A kernel static buffer is overflowed, causing a kernel panic.

### Evasion and Denial-Of-Service
Firewalls and intrusion detection systems _(IDS)_ analyse incoming datagrams using the information contained in both the datagram header and the datagram payload (TCP ports, UDP ports, SYN and ACK flags in the TCP header).

#### An attacker may use fragmentation to avoid filtering.
- Some firewalls make a decision on the first fragment and let the other fragments through by keeping track of the datagram ID.
- The first fragment of a TCP-over-IP may contain only 8 bytes (source and destination ports for both UDP and TCP). IP上的TCP的第一段可能只包含8个字节
   - Setup flags (SYN/ACK) can be “postponed” so that incoming SYN can go through.
   - Setup flags can be overwritten by using overlapping fragments. 通过修改下一段的偏移量使该数据内容可以重叠上一段的部分内容
   - In some cases, even the original src/dst port can be rewritten.

#### An attacker may use fragmentation to avoid detection.
- Some network-based IDSs do not reassemble datagrams.
- An IDS may obtain a reassembled datagram different from the one obtained at the receiving host. IDS可能会获得与在接收主机获得的数据报不同的重组数据报

#### An attacker may use fragmentation to build a DOS attack
- An attacker may exploit problems in the reassembling code (teardrop, ping of death).
- If firewalls and IDSs reassemble a datagram before analysing it, it is possible to force the system to use a large amount of memory.

**Fragrouter** is a tool that allows one to send traffic fragmented in different (creepy) ways.

### UDP Spoofing
Basically IP spoofing.
Vulnerable protocols – DNS, RPC (e.g. NFS)
1. The attacker spoofs UDP request.
2. The server responds with UDP reply to the spoofed client.

### UDP Hijacking
Builds on UDP spoofing.
1. The client sends UDP request to the server. Meanwhile, the attacker is sniffing the traffic and obtains the request.
2. The server responds with UDP reply, but it is intercepted by attacker.
3. The attacker sends spoofed UDP reply to client.

### UDP Port scan
Used to determine which UDP services are available.
1. A zero-length UDP packet is sent to each port.
2. If an **ICMP** error message “port unreachable” is received, the service is assumed to be unavailable.

Many TCP/IP stack implementations implement a limit on the error message rate, therefore this type of scan can be slow.
`nmap -sU <target IP>`

### TCP Spoofing
Aims at impersonating another host when establishing a TCP connection.

#### Mitnick attack
1. Node A trusts node B.
2. Node C wants to impersonate B with respect to A in opening a TCP connection.
3. C kills B so that B does not send annoying RST segments.
4. C sends A a TCP SYN segment in a spoofed IP packet with B’s address as the source IP and S c as the sequence number.
5. A replies with a TCP SYN/ACK segment to B with SA as the sequence number. B ignores the segment: dead or too busy.
6. C **does not receive this segment** but to finish the handshake it has to send an ACK segment with SA + 1 as the acknowledgment number.
   - C eavesdrops the SYN/ACK segment.
   - C guesses the correct sequence number.
7. Complete the 3 way handshake and install a backdoor: `rsh x-term “echo ++>> /.rhosts ”`; allows anyone to log on to Shimomura s computer.

### TCP Hijacking (Joncheray-style) 注射式攻击
Powerful technique to take control of an existing TCP connection.

#### The attacker uses spoofed TCP segments to
- insert data in the steams,
- reset an existing connection (DoS).

#### The correct sequence/acknowledgment numbers must be used.
The attacker can 
- sniff the traffic between client and server,
- guess the correct SEQ/ACK numbers.

1. The attacker waits until all the transmitted data have been acknowledged.
2. The attacker injects the data in the stream.

“Desynchronizes” the connection – by making SEQc < ACKs or SEQc > ACKs+Win (where SEQc is attacker spoofs messages from client using this SEQ number), leading to ACK storm.

#### ACK Storm
(1) The receiver of the injected data sends an acknowledgment to the apparent sender.
(2) The apparent sender replies with an Ack with the “expected” sequence number.
(3) The receiver considers this as out of sync and sends again an Ack with the “expected” sequence number.
当会话双方接收到一个不期望的数据包后，都会用自己期望的序列号返回ACK包，于是双方来回往返各自期望的序列号，形成恶性循环，最终导致ACK风暴_（仅存在于注射式会话劫持）_

Ack messages with no data are not retransmitted in case of loss. If one message is lost, the ACK storm will stop. Any attempt to communicate after stoping will generate an ACK storm.

How to deal with ACK storm?
- ARP poisoning to send Acks to non-existing HW address.
- RST packets to both parties.
- Ack storms can be blocked by the attacker using ACK packets with the right numbers.

#### This type of TCP Hijacking can completely hijack the communication channel (man-in-the 4-middle attack).
"Early desynchronization" can be achieved by the attacker by resetting existing connections and immediately opening new ones (between the same ports) with different initial sequence numbers.

#### Tools
**Hunt** is a program that allows one to monitor and hijack TCP connections.

### TCP Port scan
Used to determine the TCP services available on a victim host.

####Primitive Scanning – connect() Scan – nmap
The attacker tries to open a TCP connection to all 65535 ports of the victim host.
If the handshake is successful then the service is available.
- Advantage: no need to be root
- Disadvantage: very noisy
`nmap -sT <target IP>`

#### TCP SYN Scanning
1. The attacker sends a SYN packet.
2. If the server answers with a SYN/ACK packet then the port is open. With a RST packet, the port is closed.
The attacker sends a RST packet instead of the final ACK. The connection is never open and the event is not logged by the OS/application.
`nmap -sS <target IP>`

#### TCP FIN Scanning
1. The attacker sends a FIN-marked packet.
2. In most TCP/IP implementations
    - If the port is closed a RST packet is sent back.
    - If the port is open the FIN packet is ignored (timeout).
In _Windows_ a RST is sent back in any case, so that all ports appear to be closed.

Variation of this type of scanning
- _Xmas_ – FIN, PSH, URG set
- Null – no flags set
`nmap -sF <target IP>`

#### Idle scanning (Blind)
Uses another (mostly idle) host to “relay” the scan. 用第三方空闲主机进行端口扫描
